### PR TITLE
Add option to ignore the first events of the stylus when touching the screen.

### PR DESCRIPTION
**Pull request by hard-zero1**
Add option to ignore the first events of the stylus when touching the screen to avoid artifacts at the beginnings of strokes. The number of ignored events can be set in the preferences.

I had this problem on Kubuntu 18.04: When the stylus is pressed on the screen, the accuracy and update rate of its recognized position increases. Because of the lower precision before, when hitting the screen the cursor might jump from its previous position to the more precise point (this is especially far if the pen tip was in rapid movement right before touching), while beeing recognized as writing. This can result in artifacts that e.g. make handwritten text very hard to read. To work around this, the first events since the triggering event can be ignored.

This works great for me (when ignoring the first 3 events) and should also be a workaround for #1662 and #1821 as they seem to be the same problem.

### CODE REVIEW

File: src/gui/dialog/SettingsDialog.cpp [hunk 1]
```diff
@@ -202,6 +208,7 @@ void SettingsDialog::load() {
     loadCheckbox("cbHideVerticalScrollbar", settings->getScrollbarHideType() & SCROLLBAR_HIDE_VERTICAL);
     loadCheckbox("cbDisableScrollbarFadeout", settings->isScrollbarFadeoutDisabled());
     loadCheckbox("cbTouchWorkaround", settings->isTouchWorkaround());
+    loadCheckbox("cbIgnoreFirstStylusEvents", settings->getIgnoredStylusEvents() != 0);  // 0 means disabled, >0 enabled
```
Comment (Technius) [inline src/gui/dialog/SettingsDialog.cpp:211]: Repeated use of `settings->getIgnoredStylusEvents() != 0` is error-prone. Please store it in a local variable `const bool ignoreStylusEventsEnabled = ...` and reuse on line 227.

File: src/gui/dialog/SettingsDialog.cpp [hunk 2]
```diff
@@ -216,6 +223,13 @@ void SettingsDialog::load() {
     GtkWidget* spAutosaveTimeout = get("spAutosaveTimeout");
     gtk_spin_button_set_value(GTK_SPIN_BUTTON(spAutosaveTimeout), settings->getAutosaveTimeout());
 
+    GtkWidget* spNumIgnoredStylusEvents = get("spNumIgnoredStylusEvents");
+    if (settings->getIgnoredStylusEvents() == 0) {  // 0 means checkbox disabled, the spinButton's value should be >= 1
```
Comment (Technius) [inline src/gui/dialog/SettingsDialog.cpp:227]:
```suggestion
    if (!ignoreStylusEventsEnabled) {
```

File: src/gui/inputdevices/StylusInputHandler.cpp [hunk 3]
```diff
@@ -49,7 +55,16 @@ auto StylusInputHandler::handleImpl(InputEvent const& event) -> bool {
...
+        if (this->eventsToIgnore >= 0) {
+            if (this->eventsToIgnore == 0) {
+                this->eventsToIgnore = -1;
+                this->actionStart(event);
+            } else {
+                this->eventsToIgnore--;
+            }
+        } else {
+            this->actionMotion(event);
+        }
```
Comment (Technius) [inline src/gui/inputdevices/StylusInputHandler.cpp:67]: You can flatten the nested ifs:
```c++
if (this->eventsToIgnore == 0) {
    // ...
} else if (this->eventsToIgnore > 0) {
    // ...
} else {
    // ...
}
```

File: src/gui/inputdevices/StylusInputHandler.cpp [hunk 4]
```diff
@@ -110,7 +127,9 @@ void StylusInputHandler::setPressedState(InputEvent const& event) {
     {
         switch (event.button) {
             case 1:
-                this->deviceClassPressed = true;
+                if (!this->eventsToIgnore >= 0) {
+                    this->deviceClassPressed = true;
+                }
```
Comment (Technius) [inline src/gui/inputdevices/StylusInputHandler.cpp:132]: Looks like a typo here? As it is currently written, `!this->eventsToIgnore` is the same as `this->eventsToIgnore != 0` but then `>= 0` always evaluates to true.

**Comment by hard-zero1** [comment]
The mentioned typo was more than a typo, sorry. It should be fixed now.

File: src/gui/inputdevices/StylusInputHandler.cpp [hunk 5]
```diff
@@ -92,8 +90,10 @@ auto StylusInputHandler::handleImpl(InputEvent const& event) -> bool {
     // Trigger end of action if pen tip leaves screen or mouse button is released
     if (event.type == BUTTON_RELEASE_EVENT) {
         if (event.button == 1 || this->inputContext->getSettings()->getInputSystemTPCButtonEnabled()) {
-            if (this->inputRunning) {
+            if (this->eventsToIgnore < 0) {
                 this->actionEnd(event);
+            } else {
+                this->eventsToIgnore = -1;
```
Comment (LittleHuba) [inline src/gui/inputdevices/StylusInputHandler.cpp:96]: Should this not be done in either case?

Comment (hard-zero1) [reply src/gui/inputdevices/StylusInputHandler.cpp:96]: It is -1 already in the first case. But i can also set it every time if you want?

Comment (LittleHuba) [reply src/gui/inputdevices/StylusInputHandler.cpp:96]: Nah if it is -1 already then its good.

**Comment by Technius** [comment]
Merging now, thanks for the PR!

**Pull request merged**